From 75268a3042d7bba9551e365f34165589f3ac7472 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Mon, 22 Sep 2025 17:35:31 +0200
Subject: [PATCH 461/553] CMake: Use only working VS generators for qt_add_ui
 tests

CMake 4.2 introduced a new VS generator "Visual Studio 18 2026".

The qt_add_ui tests try to use the latest VS generator as reported by
CMake, but the CI doesn't have that VS version installed and the tests
end up failing.

Try to configure an empty project with each attempted VS version, and
only use one that actually works.

Pick-to: 6.8
Change-Id: I1b65df2c91248499d671cbd916c25496515475f4
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
(cherry picked from commit 6bd020834043c8029828a0c6dc04b876a0ceec98)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../test_qt_add_ui_common/functions.cmake     | 30 +++++++++++++++++++
 .../vs_generator_test/CMakeLists.txt          |  6 ++++
 2 files changed, 36 insertions(+)
 create mode 100644 tests/auto/cmake/test_qt_add_ui_common/vs_generator_test/CMakeLists.txt

diff --git a/tests/auto/cmake/test_qt_add_ui_common/functions.cmake b/tests/auto/cmake/test_qt_add_ui_common/functions.cmake
index b06ba25e21f..d3878da3fbc 100644
--- a/tests/auto/cmake/test_qt_add_ui_common/functions.cmake
+++ b/tests/auto/cmake/test_qt_add_ui_common/functions.cmake
@@ -8,7 +8,30 @@ function(generate_hash_folder target_name infile out_folder)
     set(${out_folder} "${short_hash}" PARENT_SCOPE)
 endfunction()
 
+function(check_generator_works out_var generator_name)
+    message(STATUS "Checking if generator '${generator_name}' works")
+    string(CONCAT source_dir
+        "${CMAKE_CURRENT_SOURCE_DIR}/../"
+        "test_qt_add_ui_common/vs_generator_test")
+    set(build_dir
+        "${CMAKE_CURRENT_BINARY_DIR}/vs_generator_test-build")
+    run_cmake_configure(SOURCE_DIR "${source_dir}"
+                        BUILD_DIR "${build_dir}"
+                        GENERATOR "${generator_name}"
+                        CLEAN_FIRST
+                        RESULT_VARIABLE cmake_result)
+
+    if("${cmake_result}" EQUAL 0)
+        set(generator_works "TRUE")
+    else()
+        set(generator_works "FALSE")
+    endif()
+    message(STATUS "Checking if generator '${generator_name}' works - ${generator_works}")
+    set(${out_var} "${generator_works}" PARENT_SCOPE)
+endfunction()
+
 function(get_latest_vs_generator output)
+    message(STATUS "Checking which latest Visual Studio generator can be used.")
     execute_process(COMMAND ${CMAKE_COMMAND} -G
         ERROR_VARIABLE CMAKE_GENERATORS_ERROR
         OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -23,7 +46,14 @@ function(get_latest_vs_generator output)
     set(last_generator "")
     foreach(generator IN LISTS vs_generators)
         string(REGEX MATCH "Visual Studio ([0-9]+) [0-9]+" unused "${generator}")
+
         if("${CMAKE_MATCH_1}" VERSION_GREATER "${last_version}")
+            # Skip Visual Studio 18 2026 because it's not installed in CI yet
+            check_generator_works(generator_works "${CMAKE_MATCH_0}")
+            if(NOT generator_works)
+                continue()
+            endif()
+
             set(last_version "${CMAKE_MATCH_1}")
             set(last_generator "${CMAKE_MATCH_0}")
         endif()
diff --git a/tests/auto/cmake/test_qt_add_ui_common/vs_generator_test/CMakeLists.txt b/tests/auto/cmake/test_qt_add_ui_common/vs_generator_test/CMakeLists.txt
new file mode 100644
index 00000000000..dbde5e369e5
--- /dev/null
+++ b/tests/auto/cmake/test_qt_add_ui_common/vs_generator_test/CMakeLists.txt
@@ -0,0 +1,6 @@
+# Copyright (C) 2025 The Qt Company Ltd.
+# SPDX-License-Identifier: BSD-3-Clause
+
+cmake_minimum_required(VERSION 3.16)
+
+project(VsGeneratorTest LANGUAGES CXX)
-- 
2.51.2

